jpx/src/main/java/io/jenetics/jpx/XMLProviderImpl.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * Java GPX Library (@__identifier__@).
+ * Copyright (c) @__year__@ Franz Wilhelmstötter
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Author:
+ *    Franz Wilhelmstötter (franz.wilhelmstoetter@gmail.com)
+ *    Antoine Vianey (https://github.com/avianey)
+ */
+package io.jenetics.jpx;
+
+import javax.xml.parsers.DocumentBuilderFactory;
+import javax.xml.stream.XMLInputFactory;
+import javax.xml.stream.XMLOutputFactory;
+
+public class XMLProviderImpl extends XMLProvider {
```

Comment:
I would like to keep the default implementation package private.
The rest looks great. 💯

Reply:
I mad eit public because some people (like me) might want a base implementation to override only one method... keeping others like the default impl...

Making it ackage private will force themn to duplicate code in that case :-/

WDYT ?

Reply:
This was the reason why I putted the default implementation into the `XMLProvider` class directly. There are two possibilities: 
1) Inheritence: put the default implementation back into the `XMLProvider`. This would than work like the Swing _Adapter_ classes.
2) Delegation: Create a static final field in the `XMLProvider` class and let implementers of provider classes use this implementation, if needed.
```java
public abstract class XMLProvider {
    public static final XMLProvider DEFAULT_PROVIDER = new XMLProviderImpl();
    ...
}
```

I'm happy with both solutions, as long as we make the default implementation itself package private ;-)

Reply:
Don't know about Swing...
Should I keep XMLProvider abstract and then have an empty XMLProviderImpl ? or remove the abstract from XMLProvider.java you proposed ?

Reply:
I think we should keep the `XMLProvider` _fully_ implemented **and** abstract. The `abstract` keyword will indicate, that an implementation is required, where the implementer can override the desired method(s). If only a different document builder is needed, just override the `documentBuilderFactory` method.
```java
public abstract class XMLProvider {
    ...
}
```
Then we can remove the _Impl_ class completely. For the default instance, we can write `new XMLProvider(){}`